[WIP] Add HomeKit compatible apps page

kimdv opened the pull request.

kimdv commented:
So contribution should not be available for the public? 

I still think we should fetch data somehow from iTunes. Otherwise we should manually keep data up to date. (If they change the name etc, app icon or so)

BalestraPatrick commented:
The contribution can be public in order for people to let us know about new HomeKit apps. Using the iTunes Search APIs would be no different than searching in the App Store directly, so it’s no use to our users. Instead, by providing them with the already known and best HomeKit apps would definitely be interesting and helpful.

kimdv commented:
So the contribution should be like the accessory. 
Only with name and link?
No more additional information ?

BalestraPatrick commented:
@kimdv yeah, it should be similar to the accessory. We should ask information such as name, link, icon picture, website and price.

BalestraPatrick commented:
Hey @kimdv, is there anything blocking this PR or may I know its status? 👍

BalestraPatrick commented:
@kimdv Sure, no worries at all! I appreciate your help, just ping me and remove the WIP status when you feel ready 😄

kimdv commented on Sources/App/Controllers/HomekitAppController.swift:
```diff
@@ -0,0 +1,149 @@
+        }
+    }
+
+    func contribute(_ req: Request) throws -> Future<View> {
```
@BalestraPatrick I was thinking of we should add contribution to the pages to `accessory/contribute` and `apps/contribute`

Be cause I think one big contribute page with a dynamic form for apps and accessories will be messy.

kimdv replied:
What do you think?

BalestraPatrick replied:
Sure, that sounds good in this case.

kimdv commented:
We also need some sort of report page here. 
Should we extend the correct report and with "Report app functionality" ?

BalestraPatrick commented:
@kimdv We can extend the usual report and simply add an option to select the app from a dropdown similar to the one we have to select the accessory now.

kimdv commented:
Should the manufactor / developer be the same? 

A lot of the manufactures of accessories develop their own app?

BalestraPatrick commented:
@kimdv Yes, I think we can have a single `publisher` data to store who is the developer and publisher.

kimdv commented:
hmm, I don't quiet follow? 
Should we have a `Publisher` table and refer to the developer?

BalestraPatrick commented:
@kimdv No, I mean that each app object should have a single publisher property of type String, to store the developer of the HomeKit app.

kimdv commented:
So no foreign key to a manufacturer or developer in case of multiple apps?

BalestraPatrick commented:
No, let's keep it simple for now :)

kimdv commented:
Hi @BalestraPatrick ! 

I think I have something that is ready for first feedback!

I have had a bit trouble about the contribute page, because if we mix it with the current, I think it gets a lot messy with a lot if app then this otherwise that, if you can follow me?

BalestraPatrick commented:
@kimdv The PR looks good, just a couple of points to modify before we can proceed to deploy this to staging for further testing:
- contribute link when there are no apps points to /contribute of accessory instead of /apps/contribute
- empty state when there are no apps is not displayed, the page is just empty.
- can we rename the table containing the apps from "homekit_apps" to simply "apps"?
- Please also change everywhere in the codebase "Homekit", to "HomeKit". A simple case-sensitive find and replace should do it..
- we should add the price to the appItem in order to let the user see the name and the price on the same line. Example: HomeDash - $4.99
- I think we should add the apps at the top in the sidebar. Before "All Accessories", we can have a single element which is "iOS HomeKit Apps" which open the page, instead of the "Apps" button at the top.
- We should add a "report issue" and "contribute" button at the bottom of the list for consistency, not at the top like the "contribute" button is now.
- In the database schema, I think we can remove the subtitle which doesn't really make sense for iOS apps.
- Why do we ask for the App Store ID in the contribution form? I don't see where it's used in the codebase. We can simply ask for the app name, price, developer name, app website, app store link and icon image link. That should be everything we need.

kimdv commented:
Regarding the "report issue" do you think something like this

![screen shot 2018-08-20 at 08 12 54](https://user-images.githubusercontent.com/5477711/44323148-ee784b00-a450-11e8-91e6-df30b958741e.png)

I think it is double with the report issue, because it is already at the bottom of the page.

kimdv commented:
It is possible to add subtitles to iOS apps as in iOS 11 App Store. 
So I think we should keep it.

BalestraPatrick commented:
@kimdv We can remove the contribute and report issue on the right if there are no apps to browse. Regarding the subtitle, I know what you mean but the subtitle is optional in the App Store. For example, if you search the Philips Hue app, you won't see any subtitle. Let's make the subtitle optional and display it in the app details page only if present.

kimdv commented:
I think I got the most now! 

I think we should reconsider the contribute at top of the page, and move it down under all the accessories like the apps. 

Or we should add a 2 step contribution. 

One pointing to apps and one to accessory

BalestraPatrick commented:
@kimdv Yeah, splitting the contribution in two pages makes sense, so we can clean up the code a bit. Can you take on that? I will do some small UI modifications to the sidebar now.

kimdv commented:
Should I add it to thi PR?

kimdv commented:
Should I add it to thi PR?

BalestraPatrick commented:
@kimdv Yes, that sounds good!

kimdv commented:
hmm, I have added a simple link. 
But we should maybe add some describing text or some images, because it is really empty that view

BalestraPatrick commented:
@kimd What page are you referring to? I just tried the project and the "report" button is now filling in the details of the accessory in the form. Try pressing the "report" button when on an "App Details" page.

kimdv commented:
@BalestraPatrick I added some performance fixes so the iOS app, manufacture and accessory counts are fetched in one request. 

Fixed the linking. But I can't reproduce the "report" issue. 
Mine forms are filled correctly. 
What view are you looking at?

BalestraPatrick commented:
@kimdv If you add an HomeKit app, and then go to its page (i.e: http://0.0.0.0:8080/apps/2) and press "Report Issue", it will go to `http://0.0.0.0:8080/report` but at this point the dropdown menu is not selected with the accessory where the user want to report the problem?
For the accessories, we pass the `id` to the report page in this way: http://0.0.0.0:8080/accessories/804/report
By doing this, when rendering the `report` page we can already pre-select the correct accessory of the issue for the user.

kimdv commented:
@BalestraPatrick fixed!
